### ISSUE TITLE
Why poststart and poststop hook are not consistent with the standard?

### ISSUE THREAD

**Issue by duguhaotian** [opencontainers/runc#1765]
```
The poststart hooks MUST be invoked by the runtime. If any poststart hook fails, the runtime MUST log a warning, but the remaining hooks and lifecycle continue as if the hook had succeeded.

The poststop hooks MUST be invoked by the runtime. If any poststop hook fails, the runtime MUST log a warning, but the remaining hooks and lifecycle continue as if the hook had succeeded.
```
[link at](https://github.com/opencontainers/runtime-spec/blob/master/runtime.md#lifecycle)

our runc implements like this

```
                        for i, hook := range c.config.Hooks.Poststart {
				if err := hook.Run(s); err != nil {
					if err := ignoreTerminateErrors(parent.terminate()); err != nil {
						logrus.Warn(err)
					}
					**return newSystemErrorWithCausef(err, "running poststart hook %d", i)**
				}
			}

                    for i, hook := range c.config.Hooks.Poststop {
			logrus.Infof("run poststop hook %d:%s", i, hook.Info())
			if err := hook.Run(s); err != nil {
				logrus.Errorf("running poststop hook %d: %s failed: %s", i, hook.Info(), err)
				**return newSystemErrorWithCausef(err, "running poststop hook %d:%s", i, hook.Info())**
			}
			logrus.Infof("poststop hook %d:%s done", i, hook.Info())
		}
```

If hook run is failed, we will return a err. But spec say:  "but the remaining hooks and lifecycle continue as if the hook had succeeded."

Why we implement like this????

Thanks!!!!

**Comment by cyphar** [comment]
There are some other problems with how hooks are implemented, see #1710 (which I think this is ultimately a dup of).

> Why we implement like this

`runc` had hook support before the specification, and the specification changed their semantics from the ones `runc` had but we haven't changed `runc` yet to follow the specification.

**Comment by wking** [comment]
> … the specification changed their semantics from the ones `runc` had…

The spec almost dropped hooks [when the create / start split landed][1] and did not update [the timing requirements][2] with sufficient clarification for a split create/start.  The current timing requirements largely landed in opencontainers/runtime-spec#532.  So it's less “the specification changed their semantics” and more “it took almost a year after hooks landed before the spec regained a detailed opinion on when the hooks should be called”.  In the interim, things were left up to runtimes, and runc went down one path with its create/start split while the spec ended up going down a different path.

But +1 to this being a dup of #1710.

[1]: https://github.com/opencontainers/runtime-spec/pull/384/files#diff-b84a8d65d8ed53f4794cd2db7e8ea731L42
[2]: https://github.com/duglin/runtime-spec/blob/be594153b522f52bebd0500cef6fe0f1d77f6a59/config.md#prestart

**Comment by duguhaotian** [comment]
Thanks for you reply
